test/libsolidity/astPropertyTests/forLoop/isSimpleCounterLoop_positive_cases.sol

```diff
@@ -0,0 +1,55 @@
+function h() pure returns (uint) {
+    return 13;
+}
+
+contract C {
+    uint[] dynArray;
+
+    function f() public {
+        /// SimplePreIncrement: isSimpleCounterLoop
+        for(uint i = 0; i < 42; ++i) {
+        }
+        /// SimplePosIncrement: isSimpleCounterLoop
+        for(int i = 0; i < 42; i++) {
+        }
+        uint x;
+        /// CounterReadLoopBody: isSimpleCounterLoop
+        for(uint i = 0; i < 42; i++) {
+            x = i;
+        }
+        /// LocalVarConditionRHS: isSimpleCounterLoop
+        for(uint i = 0; i < x; i++) {
+        }
+        uint[8] memory array;
+        /// StaticArrayLengthConditionRHS: isSimpleCounterLoop
+        for(uint i = 0; i < array.length; i++) {
+        }
+        dynArray.push();
+        /// DynamicArrayLengthConditionRHS: isSimpleCounterLoop
+        for(uint i = 0; i < dynArray.length; i++) {
+            dynArray.push(i);
+        }
```

Comment:
Oh, did not expect that one.

I guess it's safe because any expression not dependent on `i` will either become equal to `i` at some point and stop the loop or overflow before `i` does? And the fact that the length increases does not change that?

Reply:
Yeah, that's what I am assuming that happens. The code checks if the common type of the expression is the same as the left hand side and that should guarantee that either `i` catches up or the array `length` overflows.
For example, if `i` had type `uint8`, then the flag `isSimpleCounterLoop` would be `false`.

Reply:
The point in the logic for counting all these as simple loops is that the right-hand-side of the comparison does not matter at all.

Reply:
The right-hand-side can even contain `i` itself - doesn't change anything.

Reply:
As in: the only situation in which an overflow-checked increment is relevant, is if `i` has the maximal value of its type before the increment is executed. Since it doesn't change in the body, it must have had that value in the condition already. But ``type(<type of i>).max`` *never* compares ``<`` to anything. So it should be fine even if the right-hand-side of the comparison reads `i` itself.

A weird fluke would be if the right-hand-side of the expression actually *changed* the value of `i`. For that soundness actually depends on the order of evaluation of the sides of the comparison, right?

Reply:
> I.e. do we currently allow

We don't. We actually have a test against a very similar case:
```solidity
        /// CounterAssignmentConditionRHS: isSimpleCounterLoop
        for(uint i = 0; i < (i = i + 1); ++i) {
        }
```
and the check in the code is for `willBeWrittenTo` in the RHS, so it should catch `++i` as well.

Reply:
Ah yeah, I didn't see it, since the check is written very weirdly, see https://github.com/ethereum/solidity/pull/13378/files#r1368592311